queries.js

```diff
@@ -0,0 +1,234 @@
+            if (!years.find(e => e.value === year)) {
+              years.push({ value: year, label: year });
+            }
+          }
+          years.sort().reverse();
+          return years;
+        }
+      },
+      // Filter by month, in YYYY-MM- format, using regex. The event must
+      // be taking place during that month (it might surround it).
+      // Use of this filter cancels the upcoming filter
+      month: {
+        def: null,
+        async finalize() {
+          const month = query.get('month');
+
+          if (!month) {
+            return;
+          }
+          const re = new RegExp(`${month}-`, 'gi');
+
+          query.and({
+            $and: [{ startDate: re }, { endDate: re }]
+          });
+        },
+        launder: function(s) {
+          s = self.apos.launder.string(s);
+          if (!s.match(/^\d\d\d\d-\d\d$/)) {
+            return null;
+          }
+          return s;
+        },
+        async choices() {
+          const alldates = await query
+            .clone()
+            .upcoming(null)
+            .toDistinct('startDate');
+
+          const months = [{ value: null, label: 'All' }];
+          for (const eachdate of alldates) {
+            const month = eachdate.substr(0, 7);
+            if (!months.find(e => e.value === month)) {
+              months.push({ value: month, label: month });
+            }
+          }
+          months.sort().reverse();
+          return months;
+        }
+      },
+      // Filter by month, in YYYY-MM-DD format, using regex. The event must
+      // be taking place during that month (it might surround it).
+      // Use of this filter cancels the upcoming filter
+      day: {
+        def: null,
+        async finalize() {
+          const day = query.get('day');
+
+          if (!day) {
+            return;
+          }
+          const re = new RegExp(`${day}`, 'gi');
+
+          query.and({
+            $and: [{ startDate: re }, { endDate: re }]
+          });
+        },
+        launder: function(s) {
+          s = self.apos.launder.string(s);
+          if (!s.match(fullDateRegex)) {
+            return null;
+          }
+          return s;
+        },
+        async choices() {
+          const alldates = await query
+            .clone()
+            .upcoming(null)
+            .toDistinct('startDate');
+
+          const days = [{ value: null, label: 'All' }];
+          for (const eachdate of alldates) {
+            if (!days.find(e => e.value === eachdate)) {
+              days.push({ value: eachdate, label: eachdate });
+            }
+          }
+          days.sort().reverse();
+          return days;
+        }
+      },
+      // Filter for events that are active after a certain date, in YYYY-MM-DD format.
+      // The event must end on or after that day.
+      // Use of this filter cancels the upcoming filter
+      start: {
+        def: null,
+        safeFor: 'public',
+        finalize: function() {
```

Comment:
Could write `finalize() {` for brevity but it's OK.

Reply:
Agreed and done